e2e/vaultcompat/vaultcompat_test.go

```diff
@@ -0,0 +1,260 @@
+	"github.com/hashicorp/go-version"
+	nomadapi "github.com/hashicorp/nomad/api"
+	"github.com/hashicorp/nomad/testutil"
+	vaultapi "github.com/hashicorp/vault/api"
+	"github.com/shoenig/test/must"
+	"github.com/shoenig/test/wait"
+)
+
+const (
+	binDir  = "vault-bins"
+	envGate = "NOMAD_E2E_VAULTCOMPAT"
+)
+
+func TestVaultCompat(t *testing.T) {
+	if os.Getenv(envGate) != "1" {
+		t.Skip(envGate + " is not set; skipping")
+	}
+	t.Run("testVaultVersions", testVaultVersions)
+}
+
+func testVaultVersions(t *testing.T) {
+	versions := scanVaultVersions(t, getMinimumVersion(t))
+	versions.ForEach(func(b build) bool {
+		downloadVaultBuild(t, b)
+		testVaultBuild(t, b)
+		return true
+	})
+}
+
+func testVaultBuild(t *testing.T, b build) {
+	t.Run("vault("+b.Version+")", func(t *testing.T) {
+		vStop, vc := startVault(t, b)
+		defer vStop()
+		setupVault(t, vc)
+
+		nStop, nc := startNomad(t, vc)
+		defer nStop()
+		runCatJob(t, nc)
+
+		// give nomad and vault time to stop
+		defer func() { time.Sleep(5 * time.Second) }()
+	})
+}
+
+func runCatJob(t *testing.T, nc *nomadapi.Client) {
+	b, err := os.ReadFile("input/cat.hcl")
+	must.NoError(t, err)
+
+	jobs := nc.Jobs()
+	job, err := jobs.ParseHCL(string(b), true)
+	must.NoError(t, err, must.Sprint("failed to parse job HCL"))
+
+	_, _, err = jobs.Register(job, nil)
+	must.NoError(t, err, must.Sprint("failed to register job"))
+
+	must.Wait(t, wait.InitialSuccess(
+		wait.ErrorFunc(func() error {
+			allocs, _, err := jobs.Allocations(*job.ID, false, nil)
+			if err != nil {
+				return err
+			}
+			if n := len(allocs); n != 1 {
+				return fmt.Errorf("expected 1 alloc, got %d", n)
+			}
+			if s := allocs[0].ClientStatus; s != "complete" {
+				return fmt.Errorf("expected alloc status complete, got %s", s)
+			}
+			return nil
+		}),
+		wait.Timeout(20*time.Second),
+		wait.Gap(1*time.Second),
+	))
+
+	t.Log("success running cat job")
+
+	_, _, err = jobs.Deregister(*job.Name, true, nil)
+	must.NoError(t, err, must.Sprint("faild to deregister job"))
+}
+
+func startVault(t *testing.T, b build) (func(), *vaultapi.Client) {
+	path := filepath.Join(os.TempDir(), binDir, b.Version, "vault")
+	vlt := testutil.NewTestVaultFromPath(t, path)
+	return vlt.Stop, vlt.Client
+}
+
+func setupVault(t *testing.T, vc *vaultapi.Client) {
+	policy, err := os.ReadFile("input/policy.hcl")
+	must.NoError(t, err)
+
+	sys := vc.Sys()
+	must.NoError(t, sys.PutPolicy("nomad-server", string(policy)))
+
+	log := vc.Logical()
+	log.Write("auth/token/roles/nomad-cluster", role)
+
+	token := vc.Auth().Token()
+	secret, err := token.Create(&vaultapi.TokenCreateRequest{
+		Policies: []string{"nomad-server"},
+		Period:   "72h",
+		NoParent: true,
+	})
+	must.NoError(t, err, must.Sprint("failed to create vault token"))
+	must.NotNil(t, secret)
+	must.NotNil(t, secret.Auth)
+}
+
+func startNomad(t *testing.T, vc *vaultapi.Client) (func(), *nomadapi.Client) {
+	ts := testutil.NewTestServer(t, func(c *testutil.TestServerConfig) {
```

Comment:
just to log my understanding here:

`make integration-test` depends on `make dev`, so it runs that first, then this testutil func does `discover.NomadExecutable()` to find the binary that was just built, and runs that as "a real nomad agent" process.

yes?

Reply:
You got it!